### PR TITLE
Harden wrapping and table reflow (#388) (#374) (#372) (#364)

### LINKED ISSUE #364: fix(reflow): replace fragile ROW_END sentinel with structural row boundaries
## Background

Table reflow round-trips row boundaries through a magic sentinel string. In
`collect_cells` the marker is appended to a cell —
`ch.push_str(" |ROW_END|")` (`src/reflow.rs:64`) — and later `split_into_rows`
detects it with `if cell == "ROW_END"` (`src/reflow.rs:75`).

## Problem

Encoding control information inside the data stream is fragile: any table whose
cell content is literally `ROW_END` (or contains ` |ROW_END|`) will corrupt row
splitting, producing wrong output with no error. The boundary concept is also
split across two functions coupled only by an undocumented string constant.

## Objective

Carry row boundaries structurally rather than via an in-band sentinel.

- Build `Vec<Vec<String>>` directly, or
- Introduce an enum cell type (e.g. `enum Cell { Content(String), RowEnd }`) so
  boundaries are represented in the type system and cannot collide with data.

The chosen representation should make it impossible for cell content to be
mistaken for a row boundary.

## Acceptance criteria

- No magic `ROW_END` string remains in the reflow path.
- A regression test covers a table containing the literal text `ROW_END` in a
  cell and asserts correct row splitting.
- Existing reflow tests continue to pass; run the full `make` commit gate.

## References

- Requested by @leynos

### LINKED ISSUE #372: --wrap can move a footnote reference to column 1, turning it into a definition
## Summary

`--wrap` can move a footnote **reference** to the start of a line. A footnote
reference written as `... word [^96]:` (a reference immediately followed by a
colon, e.g. introducing a list) is reflowed so that `[^96]:` lands in column 1,
where CommonMark/markdownlint parse it as a footnote **definition**. This
creates a spurious duplicate definition and detaches the citation from its
sentence.

## Version

mdtablefix 0.4.0

## Reproduction

Input (`fn2.md`):

~~~markdown
The Lighthouse "Progressive Web App" category specifically checks for the technical requirements of a PWA. The audits are grouped into subcategories [^96]:

- Fast and Reliable: loads quickly.
- Installable: served over HTTPS.

[^96]: Reference 96 from the source material used to compile this guide.
~~~

Command:

```console
mdtablefix --wrap --renumber --breaks --ellipsis --fences --in-place fn2.md
```

## Actual behaviour

```diff
-The Lighthouse "Progressive Web App" category specifically checks for the technical requirements of a PWA. The audits are grouped into subcategories [^96]:
+The Lighthouse "Progressive Web App" category specifically checks for the
+technical requirements of a PWA. The audits are grouped into subcategories
+[^96]:
```

The document now contains **two** `[^96]:` lines — the reflowed reference (now
parsed as a definition) and the genuine definition — so `[^96]` is defined
twice and the in-text citation is lost.

## Expected behaviour

`--wrap` should not break a line such that a footnote reference `[^id]`
(especially when followed by `:`) becomes line-initial, since `[^id]:` at the
start of a line is a footnote definition. Keep the reference attached to the
preceding text, or avoid leaving `[^id]:` at column 1.

## Impact

The citation silently disappears from the rendered sentence and a duplicate
definition is introduced. In a larger document run through the full
`mdformat-all` pipeline, the knock-on de-duplication dropped the real
definition text entirely and left an orphaned continuation line attached to the
neighbouring footnote — a silent content loss that markdownlint does not flag.

### LINKED ISSUE #374: --wrap breaks after opening '[' of a link, producing '[ text]' (MD039 / leading space in link text)
## Summary

`mdtablefix --wrap` can place a line break **immediately after the opening `[`
of a link**, when that `[` is the last character that fits within the 80-column
limit. Because the continuation line is indented (e.g. a list-item continuation),
the link text then begins with whitespace — for example `[ trybuild]` — which:

- is a Markdown link/reference whose **link text has a leading space**, and
- trips `markdownlint` rule **MD039/no-space-in-links**, and
- changes rendered output (the visible link label gains a leading space).

The wrapper should never break between an opening `[` and its link text. It
should break **before** the `[` instead, keeping `[text]` intact on one line.

Observed with `mdtablefix 0.4.0`.

## Minimal reproducer

Input (`repro.md`) — note the `[trybuild]` reference sits right at the wrap
boundary:

```markdown
- [x] 5.1.1. Implicit fixture injection: Automatically inject fixtures when a
  step function's parameter name matches a fixture name, removing the need for
  `#[from(...)]` in most cases. [user guide][implicit-fixture-guide] ·
  [trybuild][implicit-fixture-trybuild]

[implicit-fixture-guide]: https://example.com/guide
[implicit-fixture-trybuild]: https://example.com/trybuild
```

Command:

```console
$ mdtablefix --wrap repro.md
```

Actual output (note the dangling `[` at the end of the third line, and the
indented `trybuild]` becoming ` trybuild]` once joined for rendering):

```markdown
- [x] 5.1.1. Implicit fixture injection: Automatically inject fixtures when a
  step function's parameter name matches a fixture name, removing the need for
  `#[from(...)]` in most cases. [user guide][implicit-fixture-guide] · [
  trybuild][implicit-fixture-trybuild]
```

`markdownlint` on that output (the project lints Markdown via `make markdownlint`):

```text
repro.md:6 error MD039/no-space-in-links Spaces inside link text [Context: "[ trybuild]"]
```

## Expected output

Break before the `[` so the link text is never separated from its opening
bracket:

```markdown
- [x] 5.1.1. Implicit fixture injection: Automatically inject fixtures when a
  step function's parameter name matches a fixture name, removing the need for
  `#[from(...)]` in most cases. [user guide][implicit-fixture-guide] ·
  [trybuild][implicit-fixture-trybuild]
```

## Real-world occurrences

Found in the `rstest-bdd` repository after running the project formatter
(`make fmt`, which invokes `mdtablefix --wrap ...`). Both files were
`make markdownlint`-clean before the reformat:

**1. `docs/roadmap.md:251` — MD039**

Before (correct):

```markdown
  `#[from(...)]` in most cases. [user guide][implicit-fixture-guide] ·
  [trybuild][implicit-fixture-trybuild]
```

After `mdtablefix --wrap` (broken — `[ trybuild]`):

```markdown
  `#[from(...)]` in most cases. [user guide][implicit-fixture-guide] · [
  trybuild][implicit-fixture-trybuild]
```

**2. `docs/v0-6-0-migration-guide.md:399` — MD039**

Before (correct):

```markdown
`crates/rstest-bdd-harness-gpui/tests/stateful_window.rs` and the user
guide [Stateful GPUI scenarios with durable handles][users-guide-playbook]
subsection.
```

After `mdtablefix --wrap` (broken — `[ Stateful GPUI ...]`):

```markdown
`crates/rstest-bdd-harness-gpui/tests/stateful_window.rs` and the user guide [
Stateful GPUI scenarios with durable handles][users-guide-playbook] subsection.
```

```text
docs/v0-6-0-migration-guide.md:399 error MD039/no-space-in-links Spaces inside link text [Context: "[ Stateful GPUI scenarios with..."]
```

## Environment

- `mdtablefix 0.4.0`
- Trigger flag: `--wrap` (reproduces with `--wrap` alone; in the project it runs
  as `mdtablefix --wrap --renumber --breaks --ellipsis --fences --in-place`)

### LINKED ISSUE #388: --wrap claims to support a width argument
Per docs/users-guide.md, the document says --wrap should support a width argument, but the binary clearly does not support an argument.

### LINKED ISSUE #397: Add tracing instrumentation to table-reflow row parsing and wrap link/footnote coupling
## Background

PR #395 (https://github.com/leynos/mdtablefix/pull/395) implements structural
table row boundaries (#364), protected colon-suffixed footnote references
(#372), and atomic reference-link wrapping (#374). A CodeRabbit review of that
PR flagged that none of these new decision points in `src/reflow.rs` and
`src/wrap/inline/span_helpers.rs` emit any tracing/logging, even though the
`tracing` crate is already a dependency used elsewhere in the codebase (see
issues #305, #306, #296).

Raised by @leynos via
https://github.com/leynos/mdtablefix/pull/395#issuecomment (CodeRabbit review
discussion on the failed-checks summary).

## Problem

The reflow and wrap pipelines silently make structural decisions — where a
table row boundary falls, whether a footnote reference gets coupled to a
trailing colon, whether a reference-link label is kept atomic — with no
observability. Diagnosing regressions in these paths (e.g. a
misparsed row, an unexpectedly split link) currently requires attaching a
debugger or adding ad hoc `eprintln!` statements.

## Objective

Add structured `tracing::trace!`/`debug!` events at the key decision points
introduced by #364, #372 and #374:

- `parse_rows` in `src/reflow.rs`: emit an event per parsed row (or on
  discarding an empty row) with fields such as `row_index` and `cell_count`.
- `should_couple_whitespace` / `try_couple_footnote_reference` in
  `src/wrap/inline/span_helpers.rs`: emit an event when a footnote reference
  is coupled to a following colon, or when a reference-link label is kept
  atomic, with fields identifying the span kind and token content.
- Any failure/fallback boundaries in these paths (e.g. unterminated brackets)
  should emit an event with a stable `error_category`-style field.

Follow the existing tracing conventions used in
`src/wrap/tokenize/parsing.rs` (see issue #306) for event naming and field
style.

## Acceptance criteria

- New `trace!`/`debug!` events exist at the row-parsing and
  link/footnote-coupling decision points described above.
- At least one `#[traced_test]` integration test per new instrumentation
  point asserts the event fires via `logs_contain(...)`.
- No behavioural change to reflow or wrap output; instrumentation only.
- `make check-fmt`, `make lint`, `make typecheck`, and `make test` pass.

## References

- PR #395: https://github.com/leynos/mdtablefix/pull/395
- Issue #364, #372, #374
- Related tracing issues: #305, #306, #296

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch keeps `--wrap` as a parameterless 80-column flag and corrects the documentation that previously implied it accepted a width. It also keeps reference-style links and colon-suffixed footnote references attached across wrap boundaries, and replaces the table reflow `ROW_END` sentinel with structural row boundaries.

Closes #388.
Closes #374.
Closes #372.
Closes #364.
Closes #397.

## Review walkthrough

- Start with [`src/main.rs`](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/src/main.rs), [`src/process.rs`](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/src/process.rs), and the [CLI regression tests](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/tests/wrap/cli.rs) for the parameterless flag contract, fixed 80-column pipeline, and assigned-value rejection.
- Then review [`src/wrap/tokenize/parsing.rs`](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/src/wrap/tokenize/parsing.rs), [`src/wrap/inline.rs`](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/src/wrap/inline.rs), and [`src/wrap/inline/span_helpers.rs`](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/src/wrap/inline/span_helpers.rs) for reference-link tokenization and footnote attachment.
- Review [`src/reflow.rs`](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/src/reflow.rs) and [ADR 0001](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/docs/adrs/0001-table-reflow-pipeline.md) for structural row boundaries.
- Finish with the [README](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/README.md), [user's guide](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/docs/users-guide.md), [developer's guide](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/docs/developers-guide.md), and [changelog](https://github.com/leynos/mdtablefix/blob/51f6c7673c43461d3ae3642fa0a7a23a7a990a99/CHANGELOG.md) for the documented behaviour and all four issue outcomes.

## Validation

- `git diff --check`: passed.
- `make check-fmt`: passed.
- `make lint`: passed.
- `make typecheck`: passed.
- `make test`: passed.
- `make markdownlint`: passed.
- `make nixie`: passed.
- `mbake validate Makefile`: passed.
- `coderabbit review --agent`: passed with zero findings across the final 17-file branch review.

## Notes

The `--wrap` option deliberately accepts no parameter and always uses the public `WRAP_COLS` value of 80 display columns. No roadmap task or ExecPlan was identified for this issue bundle.

## References

- [Lody session](https://lody.ai/leynos/sessions/5366d71c-8fa1-4cff-bdb1-464fc728710f)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,9 @@
 
 ### Added
 
+- Allow `--wrap=<width>` to accept a positive display-column width while bare
+  `--wrap` retains the 80-column default.
+  ([#388](https://github.com/leynos/mdtablefix/issues/388))
```
[the remark this answers is not in the document]
Comment (leynos) [reply CHANGELOG.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/wrap/cli.rs [hunk 2]
```diff
@@ -28,6 +28,29 @@ fn test_cli_wrap_option() -> Result<(), Box<dyn std::error::Error>> {
...
+    Ok(())
+}
+
+#[test]
+fn test_cli_wrap_rejects_zero_width() -> Result<(), Box<dyn std::error::Error>> {
+    run_cli_with_stdin(&["--wrap=0"], "alpha beta\n")?
+        .failure()
+        .stderr(predicates::str::contains("invalid value '0'"));
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/wrap/cli.rs:37]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply CHANGELOG.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/wrap/cli.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (leynos) [reply tests/wrap/cli.rs:37]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat documentation and validation coverage as in scope).

### ❌ Failed checks (1 error, 3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | No file-backed regression proves `--wrap FILE` stays positional; all new wrap tests use stdin or in-place, so the equals-sign guarantee can still regress. | Add a temp-file CLI regression that runs `mdtablefix --wrap <file>` and asserts the file is processed, not rejected as an invalid width. |
| Testing (Unit And Behavioural) | ⚠️ Warning | CLI tests cover `--wrap=20` and `--wrap=0`, but no end-to-end test proves `--wrap FILE` keeps `FILE` positional; that edge case stays unverified. | Add a binary-level regression test that runs `mdtablefix --wrap <tempfile>` and asserts the file is processed, not parsed as a width value. |
| Testing (Property / Proof) | ⚠️ Warning | The PR introduces broad invariants, but the new coverage is example-based only; no new property/proof tests exercise the added cases. | Add proptest cases for custom wrap-width parsing, reference-link/`[^id]:` atomicity, and table splitting with arbitrary cell content. |
| Observability | ⚠️ Warning | No new logging, metrics, or tracing was added in the changed wrap/reflow paths; the only existing tracing remains in unrelated token parsing helpers. | Add structured trace/debug events at wrap-width selection, row parsing, and failure boundaries, with stable fields like file, wrap\_width, and error category. |

**Comment by pandalump** [comment]
@coderabbitai resume

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply CHANGELOG.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/reflow.rs [hunk 3]
```diff
@@ -41,49 +39,80 @@ const LEADING_EMPTY_CELL_MARKER: &str = "\u{1d}";
...
     }
-    if !current.is_empty() {
-        rows.push(current);
+
+    chunks.push(&raw[chunk_start..]);
+    chunks
+}
+
+fn retain_parsed_row(row_index: usize, row: &[String]) -> bool {
+    if row.iter().all(String::is_empty) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/reflow.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/reflow.rs (1)

`78-95`: _🗄️ Data Integrity & Integration_ | _🟠 Major_
**Disambiguate empty-cell delimiters from row boundaries.**
`split_row_chunks` treats `|` followed by whitespace and another `|` as a row boundary, although that is also valid empty-cell syntax. Inputs such as `| A |  | C |`, or rows ending with an empty cell before the next input line, lose cells. Preserve original line-boundary metadata or parse rows and cells in one stateful pass.
- `src/reflow.rs#L78-L95`: stop inferring boundaries from delimiter spacing alone.
- `src/reflow/tests.rs#L29-L47`: allow the final generated cell to be empty.
- `src/reflow/tests.rs#L88-L99`: add a whitespace-padded empty-cell regression.
- `src/reflow/tests.rs#L125-L157`: run the property over padded and trailing-empty rows.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/reflow.rs` around lines 78 - 95, Update split_row_chunks in src/reflow.rs
(anchor 78-95) to preserve original line-boundary metadata or parse rows and
cells in one stateful pass, so whitespace-padded “|  |” delimiters remain empty
cells and are not mistaken for row boundaries. In src/reflow/tests.rs (29-47),
allow the final generated cell to be empty; at 88-99, add regression coverage
for whitespace-padded empty cells; and at 125-157, extend the property cases to
include padded rows and trailing-empty rows.

_Source: Coding guidelines_

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat documentation and validation coverage as in scope).

### ❌ Failed checks (2 errors, 1 warning, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The new row-splitting tests only round-trip already-separated rows, so they never exercise `split_within_line == true` or a marker-only row; the empty-row/split bug can still slip through. | Add a `parse_rows` regression with a structurally split row and a row of only protected empties, and assert the cleaned result drops the empty row. |
| Security And Privacy | ❌ Error | New tracing emits raw table rows and token text via `ret`/`token` fields, so user document content can leak into debug/trace logs. | Drop `ret` on content-bearing instruments and redact token/row values to metadata only (indices, kinds, error categories, hashes if needed). |
| Out of Scope Changes check | ⚠️ Warning | A new date-sequence property-test module is added without any linked issue, which is unrelated to the wrapping and table-reflow scope. | Move the date-matching property tests into a separate PR, or tie them to a linked issue if they are required here. |

**Comment by leynos** [comment]
@coderabbitai resume

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/wrap/tokenize/parsing.rs (1)

`316-324`: _📐 Maintainability & Code Quality_ | _🟠 Major_ | _⚡ Quick win_
**Convert the new tests to `rstest`.** Use `#[rstest]` for `src/wrap/tokenize/parsing.rs#L316-L324` and extract the repeated token setup in `src/wrap/inline/span_helper_tracing_tests.rs#L44-L102` into fixtures and parameterised cases.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/wrap/tokenize/parsing.rs` around lines 316 - 324, Convert the test at
src/wrap/tokenize/parsing.rs lines 316-324 to use rstest, replacing the plain
test attribute while preserving its assertions. In
src/wrap/inline/span_helper_tracing_tests.rs lines 44-102, extract repeated
token setup into rstest fixtures and express the related cases as parameterized
tests; update every affected case there accordingly.

_Source: Coding guidelines_

**Comment by leynos** [comment]
@coderabbitai resume

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

Developer Documentation ⚠️ Warning docs/developers-guide.md covers the reflow ADR, but it still omits row\_parsing helpers and parse\_rows trace fields (row\_index/cell\_count/error\_category), so the new boundary isn’t fully documented. Document the new row\_parsing module/helpers and parse\_rows trace fields in the developer guide, then sync any related ADR/roadmap/ExecPlan entries if present.
